flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java

```diff
@@ -98,9 +98,16 @@ public RecordsWithSplitIds<ConsumerRecord<byte[], byte[]>> fetch() throws IOExce
         ConsumerRecords<byte[], byte[]> consumerRecords;
         try {
             consumerRecords = consumer.poll(Duration.ofMillis(POLL_TIMEOUT));
-        } catch (WakeupException we) {
-            return new KafkaPartitionSplitRecords(
-                    ConsumerRecords.empty(), kafkaSourceReaderMetrics);
+        } catch (WakeupException | IllegalStateException e) {
+            // IllegalStateException will be thrown if the consumer is not assigned any partitions.
+            // This happens if all assigned partitions are invalid or empty (starting offset >=
```

Comment:
Makes sense. Would it be even better to avoid sending the split from the enumerator if the split is invalid?

Reply:
Thanks for the review! I think it's by-design that some kinds of starting / stopping offsets like earliest, latest and committed are handled on `SourceReader` (see `ReaderHandledOffsetsInitializer`) to avoid some race conditions, like the earliest offset changes because of log deletion when split is delivered from enumerator on JM to reader on TM.

Reply:
The design makes sense, but I was thinking in terms of the specified offsets initializer. Looks like it is handled properly by the enumerator in that case though.